api/pkg/transformer/feast/feature_retriever_test.go

```diff
@@ -2654,3 +2656,107 @@ func Benchmark_buildEntitiesRequest_geohashArrays(b *testing.B) {
 		_ = request
 	}
 }
+
+var (
+	defaultMockFeastResponse = &feast.OnlineFeaturesResponse{
+		RawResponse: &serving.GetOnlineFeaturesResponse{
+			FieldValues: []*serving.GetOnlineFeaturesResponse_FieldValues{
+				{
+					Fields: map[string]*feastTypes.Value{
+						"driver_trips:average_daily_rides": feast.DoubleVal(1.1),
+						"driver_id":                        feast.StrVal("1001"),
+					},
+					Statuses: map[string]serving.GetOnlineFeaturesResponse_FieldStatus{
+						"driver_trips:average_daily_rides": serving.GetOnlineFeaturesResponse_PRESENT,
+						"driver_id":                        serving.GetOnlineFeaturesResponse_PRESENT,
+					},
+				},
+			},
+		},
+	}
+
+	defaultFeatureTableSpecs = []*spec.FeatureTable{
+		{
+			Project: "default",
+			Entities: []*spec.Entity{
+
+				{
+					Name:      "driver_id",
+					ValueType: "STRING",
+					Extractor: &spec.Entity_JsonPath{
+						JsonPath: "$.driver_id",
+					},
+				},
+			},
+			Features: []*spec.Feature{
+				{
+					Name:         "driver_trips:average_daily_rides",
+					DefaultValue: "0.0",
+					ValueType:    "DOUBLE",
+				},
+			},
+		},
+	}
+)
+
+func TestFeatureRetriever_RetriesRetrieveFeatures_MaxConcurrent(t *testing.T) {
+	mockFeast := &mocks.Client{}
+
+	for i := 0; i < 3; i++ {
+		mockFeast.On("GetOnlineFeatures", mock.Anything, mock.Anything).
+			Return(defaultMockFeastResponse, nil).
+			Run(func(arg mock.Arguments) {
+				time.Sleep(500 * time.Millisecond)
+			})
+	}
+
+	compiledJSONPaths, err := CompileJSONPaths(defaultFeatureTableSpecs)
+	assert.NoError(t, err)
+
+	compiledExpressions, err := CompileExpressions(defaultFeatureTableSpecs)
+	assert.NoError(t, err)
+
+	jsonPathStorage := jsonpath.NewStorage()
+	jsonPathStorage.AddAll(compiledJSONPaths)
+	expressionStorage := expression.NewStorage()
+	expressionStorage.AddAll(compiledExpressions)
+	entityExtractor := NewEntityExtractor(jsonPathStorage, expressionStorage)
+
+	options := &Options{
+		BatchSize: 100,
+
+		FeastTimeout:                     1 * time.Second,
+		FeastClientMaxConcurrentRequests: 2,
+		FeastClientSleepWindow:           0,
+	}
+
+	logger, _ := zap.NewDevelopment()
+
+	var requestJson transTypes.JSONObject
+	err = json.Unmarshal([]byte(`{"driver_id":"1001"}`), &requestJson)
+	assert.NoError(t, err)
+
+	fr := NewFeastRetriever(mockFeast, entityExtractor, defaultFeatureTableSpecs, options, nil, logger)
+
+	var good, bad uint32
+
+	for i := 0; i < 3; i++ {
+		go func() {
+			_, err := fr.RetrieveFeatureOfEntityInRequest(context.Background(), requestJson)
+			if err != nil {
+				atomic.AddUint32(&bad, 1)
+			} else {
+				atomic.AddUint32(&good, 1)
+			}
+		}()
+
+		time.Sleep(10 * time.Millisecond)
+	}
+
+	time.Sleep(2 * time.Second)
+
+	assert.GreaterOrEqual(t, atomic.LoadUint32(&bad), uint32(0))
```

Comment:
It seems the check is too lenient, is it possible to just use equal?

Reply:
Updated the implementation to be more consistent. Previously, we use the error channel from Hystrix's GoC function. Now we use fallback function that sends the error to batchResultChan. Please refer to: https://github.com/gojek/merlin/blob/transformer-hystrix/api/pkg/transformer/feast/feature_retriever.go#L417-L420